tests/test_fee_burner.py

```diff
@@ -0,0 +1,377 @@
+
+
+DEAD_ADDRESS = '0x000000000000000000000000000000000000dEaD'
+
+
+def mint_token(w3, token, owner, users):
+    for user in users:
+        tx_hash = token.functions.mint(user, HUGE_AMOUNT).transact()
+        w3.eth.waitForTransactionReceipt(tx_hash)
+
+    w3.eth.waitForTransactionReceipt(
+        token.functions.finishMinting().transact({'from': owner})
+    )
+
+
+@pytest.fixture()
+def omg_token(w3, operator, accounts):
+    # NOTE: default account is set to the operator, so it automagically deploys as if deployed by the operator
+    contract_interface = {'abi': OMGTOKEN_CONTRACT_ABI, 'bin': OMGTOKEN_CONTRACT_BYTECODE}
+    address = deploy_contract(w3, contract_interface)
+
+    omg_token_contract = w3.eth.contract(
+        address=address,
+        abi=contract_interface['abi']
+    )
+
+    mint_token(w3, omg_token_contract, operator, accounts[0:3])
+
+    w3.eth.waitForTransactionReceipt(
+        omg_token_contract.functions.transferOwnership(DEAD_ADDRESS).transact()
+    )
+
+    return omg_token_contract
+
+
+@pytest.fixture()
+def other_token(w3, operator, accounts):
+    compiled_erc20_token = compile_source_file('./tests/sample_erc20_contract/MintableToken.sol')
+    contract_interface = compiled_erc20_token['./tests/sample_erc20_contract/MintableToken.sol:MintableToken']
+
+    address = deploy_contract(w3, contract_interface)
+
+    erc20_token = w3.eth.contract(
+        address=address,
+        abi=contract_interface['abi']
+    )
+
+    mint_token(w3, erc20_token, operator, accounts[0:3])
+
+    w3.eth.waitForTransactionReceipt(
+        erc20_token.functions.transferOwnership(DEAD_ADDRESS).transact()
+    )
+
+    return erc20_token
+
+
+@pytest.fixture()
+def fee_burner(w3, provider, omg_token):
+    compiled_fee_burner = compile_source_file('./contracts/FeeBurner.sol')
+    contract_interface = compiled_fee_burner['./contracts/FeeBurner.sol:FeeBurner']
+
+    address = deploy_contract(w3, contract_interface, omg_token.address)
+    fee_burner_contract = w3.eth.contract(
+        address=address,
+        abi=contract_interface['abi']
+    )
+
+    fee_burner_contract.transact().addSupportFor(ZERO_ADDRESS, 1, 1)
+    mine_blocks(provider, fee_burner_contract.call().NEW_RATE_MATURITY_MARGIN() + 1)
+
+    return fee_burner_contract
+
+
+# TESTS
+
+def test_add_support_for_some_token(fee_burner, other_token, operator):
+    # when: an operator adds support for some token
+    fee_burner.functions.addSupportFor(other_token.address, 1, 123).transact()
+
+    # then: this token is supported
+    assert fee_burner.functions.getExchangeRate(other_token.address).call()[-2:] == [1, 123]
+
+
+def test_add_support_for_a_token_by_a_non_operator(fee_burner, non_operator, other_token):
+    # when: a non operator tries to add support for a token
+    # then: an error occurs
+    with pytest.raises(ValueError):  # TODO: now ValueError is thrown instead of TransactionFailed
```

Comment:
this is a little bit suspicious, but maybe that's what w3 throws on error in contract execution. Expecting a specific `TransactionFailed` seemed safer...

Reply:
`TransactionFailed` is purely `eth-tester`'s exception and I do not use this package. This is what `web3` throws. A workaround could be to use the proposed wrapper (d060b75dbbb980f1b0b2254a38339c1bf5c14a4f) of the contract and in its implementation I would catch `ValueError` and throw `TransactionFailed` exception.

Reply:
the point is that `ValueError` is too generic, and in tests you'd like to be sure that you're action is throwing a specific exception that indicates that the transaction failed. I think that in `ValueError` there's a msg field that elaborates on the reason of failure, that maybe you could check. You'll then have to write your own `with` syntax to wrap around `with py.test raises`.

ah, I remember @4gn3s did it once for a former project, there it is:

```
class RaisesTransactionFailed:
    # see https://github.com/pipermerriam/web3.py/issues/75
    def __enter__(self):
        return self

    def __exit__(self, exception_type, exception_value, traceback):
        with pytest.raises(TransactionFailed):
            if exception_type == ValueError and exception_value[0]['data']['type'] == u'TransactionFailed':
                raise TransactionFailed
        return True
```

then
```
       with RaisesTransactionFailed():
            contract.transact()...
```